File: app/src/main/java/com/locallinkonline/stcatherineschool/view/LunchViewModel.java
package com.locallinkonline.stcatherineschool.view;

import android.app.Application;
import android.text.format.DateUtils;

import com.locallinkonline.locallinkschool.view.LiveDataViewModel;
import com.locallinkonline.stcatherineschool.room.entity.LunchEntity;
import com.locallinkonline.stcatherineschool.room.repository.StCatherineRepository;

import java.util.List;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

public class LunchViewModel extends LiveDataViewModel<List<LunchEntity>> {

    private final StCatherineRepository stCatherineRepository;

    private static long lastUpdate = System.currentTimeMillis();

    public LunchViewModel(Application application) {
        super(application);
        stCatherineRepository = new StCatherineRepository(application);
        data = stCatherineRepository.getLunches();
    }

    public LiveData<List<LunchEntity>> getData() {

        long millisSinceLastUpdate = System.currentTimeMillis() - lastUpdate;

        if(millisSinceLastUpdate > 5 * DateUtils.MINUTE_IN_MILLIS) {
            stCatherineRepository.checkForNewLunches();
            lastUpdate = System.currentTimeMillis();
        }
        return data;
    }
}

File: app/src/main/java/com/locallinkonline/stcatherineschool/view/LunchViewModel.kt
package com.locallinkonline.stcatherineschool.view

import android.app.Application
import android.text.format.DateUtils

import com.locallinkonline.locallinkschool.view.LiveDataViewModel
import com.locallinkonline.stcatherineschool.room.entity.LunchEntity
import com.locallinkonline.stcatherineschool.room.repository.StCatherineRepository

import androidx.lifecycle.LiveData
import com.locallinkonline.stcatherineschool.rest.api.StCatherineAsync

class LunchViewModel(application: Application) : LiveDataViewModel<List<LunchEntity>>(application) {

    private val stCatherineRepository: StCatherineRepository

    init {
        stCatherineRepository = StCatherineRepository(application)
        liveData = stCatherineRepository.lunches
    }

    override fun getData(): LiveData<List<LunchEntity>>? {

        val millisSinceLastUpdate = System.currentTimeMillis() - lastUpdate

        if (this.liveData!!.value == null || this.liveData!!.value!!.isEmpty() || millisSinceLastUpdate > 5 * DateUtils.MINUTE_IN_MILLIS) {
            StCatherineAsync.updateLunches()
            lastUpdate = System.currentTimeMillis()
        }
        return liveData
    }

    companion object {

        private var lastUpdate = System.currentTimeMillis()
    }
}
